extmod/vfs_littlefs.c

```diff
@@ -0,0 +1,521 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "py/runtime.h"
+#include "py/stream.h"
+#include "py/binary.h"
+#include "py/objarray.h"
+#include "py/mperrno.h"
+#include "extmod/vfs.h"
+#include "extmod/vfs_littlefs.h"
+
+#if MICROPY_VFS && MICROPY_VFS_LITTLEFS
+
+size_t strspn(const char *s, const char *accept) {
+    const char *ss = s;
+    while (*s && strchr(accept, *s) != NULL) {
+        ++s;
+    }
+    return s - ss;
+}
+
+size_t strcspn(const char *s, const char *reject) {
+    const char *ss = s;
+    while (*s && strchr(reject, *s) == NULL) {
+        ++s;
+    }
+    return s - ss;
+}
+
+STATIC int dev_ioctl(const struct lfs_config *c, int cmd, int arg) {
+    mp_obj_t dest[4];
+    mp_load_method(MP_OBJ_FROM_PTR(c->context), MP_QSTR_ioctl, dest);
+    dest[2] = MP_OBJ_NEW_SMALL_INT(cmd);
+    dest[3] = MP_OBJ_NEW_SMALL_INT(arg);
+    return mp_obj_get_int(mp_call_method_n_kw(2, 0, dest));
+}
+
+STATIC int dev_read(const struct lfs_config *c, lfs_block_t block, lfs_off_t off, void *buffer, lfs_size_t size) {
+    mp_obj_vfs_littlefs_t *vfs = c->context;
+
+    mp_obj_array_t ar = {{&mp_type_bytearray}, BYTEARRAY_TYPECODE, 0, size, buffer};
+
+    mp_obj_t dest[5];
+    mp_load_method(MP_OBJ_FROM_PTR(vfs), MP_QSTR_readblocks, dest);
+    dest[2] = MP_OBJ_NEW_SMALL_INT(block);
+    dest[3] = MP_OBJ_NEW_SMALL_INT(off);
+    dest[4] = MP_OBJ_FROM_PTR(&ar);
+    mp_call_method_n_kw(3, 0, dest);
+
+    // TODO handle error return
+
+    return 0;
+}
+
+STATIC int dev_prog(const struct lfs_config *c, lfs_block_t block, lfs_off_t off, const void *buffer, lfs_size_t size) {
+    mp_obj_vfs_littlefs_t *vfs = c->context;
+
+    mp_obj_array_t ar = {{&mp_type_bytearray}, BYTEARRAY_TYPECODE, 0, size, (void*)buffer};
+
+    mp_obj_t dest[5];
+    mp_load_method(MP_OBJ_FROM_PTR(vfs), MP_QSTR_writeblocks, dest);
+    dest[2] = MP_OBJ_NEW_SMALL_INT(block);
+    dest[3] = MP_OBJ_NEW_SMALL_INT(off);
+    dest[4] = MP_OBJ_FROM_PTR(&ar);
+    mp_call_method_n_kw(3, 0, dest);
+
+    return 0;
+}
+
+STATIC int dev_erase(const struct lfs_config *c, lfs_block_t block) {
+    dev_ioctl(c, 4, block); // erase
+    // TODO handle error return
+    return 0;
+}
+
+STATIC int dev_sync(const struct lfs_config *c) {
+    dev_ioctl(c, 0, 0); // sync
+    // TODO handle error return
+    return 0;
+}
+#if (LFS_VERSION >= 0x00020000)
+static uint8_t __attribute__ ((aligned (64))) lookahead_buffer[128/8];
```

Comment:
@andrewleech I'm pretty sure this should be `lookahead_buffer[128]`...!  LFS2 changed this, compared to LFS1.  Having a small buffer will lead to memory corruption.

Reply:
Ah yes, here's the malloc for that buffer when lfs is used with dynamic memory. 
https://github.com/ARMmbed/mbed-littlefs/blob/master/littlefs/lfs2.c#L3218
It doesn't have the divide by 8, you're right. That'll be why I had some corruption when I first tested LFSv2!